plugin/dnstap/dnstapio/io.go

```diff
@@ -2,67 +2,99 @@ package dnstapio
 
 import (
 	"log"
+	"net"
+	"time"
 
 	tap "github.com/dnstap/golang-dnstap"
+	fs "github.com/farsightsec/golang-framestream"
 	"github.com/golang/protobuf/proto"
 )
 
-// DnstapIO wraps the dnstap I/O routine.
-type DnstapIO struct {
-	protocol Protocol
-	queue    chan tap.Dnstap
-	stop     chan bool
+const (
+	tcpTimeout   = 4 * time.Second
+	flushTimeout = 1 * time.Second
+	queueSize    = 1000
+)
+
+type dnstapIO struct {
+	enc   *fs.Encoder
+	conn  net.Conn
+	queue chan tap.Dnstap
 }
 
-// Protocol is either `out.TCP` or `out.Socket`.
-type Protocol interface {
-	// Write takes a single frame at once.
-	Write([]byte) (int, error)
+// New dnstap I/O routine.
+func New() DnstapIO {
+	return &dnstapIO{queue: make(chan tap.Dnstap, queueSize)}
+}
 
-	Close() error
+// DnstapIO interface
+type DnstapIO interface {
+	Connect(endpoint string, socket bool) error
+	Dnstap(payload tap.Dnstap)
+	Close()
 }
 
-// New dnstap I/O routine from Protocol.
-func New(w Protocol) *DnstapIO {
-	dio := DnstapIO{}
-	dio.protocol = w
-	dio.queue = make(chan tap.Dnstap, 10)
-	dio.stop = make(chan bool)
+// Connect to the Dnstap endpoint
+func (dio *dnstapIO) Connect(endpoint string, socket bool) error {
+	var err error
+	if socket {
+		dio.conn, err = net.Dial("unix", endpoint)
+	} else {
+		dio.conn, err = net.DialTimeout("tcp", endpoint, tcpTimeout)
+	}
+	if err != nil {
+		return err
+	}
+	dio.enc, err = fs.NewEncoder(dio.conn, &fs.EncoderOptions{
+		ContentType:   []byte("protobuf:dnstap.Dnstap"),
+		Bidirectional: true,
+	})
+	if err != nil {
+		return err
+	}
 	go dio.serve()
-	return &dio
+	return nil
 }
 
 // Dnstap enqueues the payload for log.
-func (dio *DnstapIO) Dnstap(payload tap.Dnstap) {
+func (dio *dnstapIO) Dnstap(payload tap.Dnstap) {
 	select {
 	case dio.queue <- payload:
 	default:
-		log.Println("[WARN] Dnstap payload dropped.")
+		log.Printf("[ERROR] Dnstap payload dropped")
 	}
 }
 
-func (dio *DnstapIO) serve() {
+// Close waits until the I/O routine is finished to return.
+func (dio *dnstapIO) Close() {
+	close(dio.queue)
+	dio.conn.Close()
+}
+
+func (dio *dnstapIO) serve() {
+	timeout := time.After(flushTimeout)
```

Comment:
make this a time.Ticker? Resolves the need to reassign to timeout ?

Reply:
There is a concern - if Flush() takes more time than ticker interval - we have useless pending Flush() call(s).